libraries/anvill_passes/src/ConvertXorToCmp.cpp

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright (c) 2021 Trail of Bits, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <anvill/ABI.h>
+#include <glog/logging.h>
+#include <llvm/IR/Constants.h>
+#include <llvm/IR/IRBuilder.h>
+#include <llvm/IR/InstIterator.h>
+#include <llvm/IR/Instructions.h>
+#include <llvm/IR/Module.h>
+#include <llvm/Pass.h>
+
+#include <vector>
+
+namespace anvill {
+namespace {
+
+class ConvertXorToCmp final : public llvm::FunctionPass {
+ public:
+  ConvertXorToCmp(void) : llvm::FunctionPass(ID) {}
+
+  bool runOnFunction(llvm::Function &func) override;
+
+ private:
+  static char ID;
+};
+
+char ConvertXorToCmp::ID = '\0';
+
+// Get which operand for a binary operator is an ICmpInst
+// nullptr if neither
+static llvm::ICmpInst *getComparisonOperand(llvm::BinaryOperator *op) {
+  auto rhs_cmp = llvm::dyn_cast<llvm::ICmpInst>(op->getOperand(1));
+  if (rhs_cmp) {
+    return rhs_cmp;
+  }
+
+  auto lhs_cmp = llvm::dyn_cast<llvm::ICmpInst>(op->getOperand(0));
+  if (lhs_cmp) {
+    return lhs_cmp;
+  }
+
+  return nullptr;
+}
+
+static llvm::Value *negateCmpPredicate(llvm::ICmpInst *cmp) {
+  auto pred = cmp->getPredicate();
+  llvm::IRBuilder<> ir(cmp);
+  llvm::ICmpInst::Predicate new_pred = llvm::CmpInst::getInversePredicate(pred);
+
+  // create a new compare with negated predicate
+  return ir.CreateICmp(new_pred, cmp->getOperand(0), cmp->getOperand(1));
+}
+
+bool ConvertXorToCmp::runOnFunction(llvm::Function &func) {
+  std::vector<llvm::BinaryOperator *> xors;
+
+  for (auto &inst : llvm::instructions(func)) {
+
+    // check for binary op
+    if (auto binop = llvm::dyn_cast<llvm::BinaryOperator>(&inst)) {
+
+      // binary op is a xor
+      if (binop->getOpcode() == llvm::Instruction::Xor) {
+
```

Comment:
Call `getComparisonOperand` here, otherwise your later `DLOG(ERROR)` isn't really an error?

Reply:
So I purposely do not call getComparisonOperand here since I also need the non-comparison operand. I think a more reasonable solution may be to modify getComparisonOperand() to return both operands, since the logic is nearly identical

Reply:
reworked getComparisonOperand => getComparisonOperands